### PR TITLE
Parser exp PR review

### PR DESCRIPTION
**Pull request by Jimobob**

### CODE REVIEW

**Comment by rmh2805** [comment]
Have we done unit testing on the parser branch?

File: src/parseSet.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+class parseSet {
+
+    private static final String DELIMITER = "[^a-zA-Z0-9_,]+";
+    private static final String FIRSTPATH = "src/FIRST.txt";
+    private static final String PREDICTPATH = "src/PREDICT.txt";
+    static Map<String, Map<String, Boolean>> FIRST = new HashMap<>();
+    static Map<String, Map<String, List<String>>> PREDICT = new HashMap<>();
+
+    /**
+     * Populate FIRST or PREDICT data structure for LL parsing given file with
```
Comment (rmh2805) [inline src/parseSet.java]: (For the whole file) Why is this in its own file, as opposed to in the parse file?

File: src/parseTree/nodes/node.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package src.parseTree.nodes;
+
+public interface node {
+    public void addChild(Object child);
+}
```
Comment (rmh2805) [inline src/parseTree/nodes/node.java]: If we're doing this, and I can understand the appeal, there should also be a verify statement in there, so that each node can focus on verifying itself to save complexity in the long run

File: src/parseTree/nodes/asmt.java [hunk 3]
```diff
@@ -1,23 +1,21 @@
...
-import src.parseTree.tokens.id;
+import java.util.ArrayList;
+import java.util.List;
 
-public class asmt extends stmt<Integer> {
-    private type t;
-    private id name;
-    private asmt_op op;
-    private expr exp;
-    private end_stmt endStmt;
```
Comment (rmh2805) [inline src/parseTree/nodes/asmt.java]: Why did you remove the categories for ONLY some of the nodes and not all?

File: src/parseTree/nodes/charAt_expr.java [hunk 4]
```diff
@@ -1,26 +1,13 @@
...
+import java.util.ArrayList;
+import java.util.List;
 
-public class charAt_expr extends str_expr {
-    charAt_label op;
-    start_paren startParen;
-    str_val strExpr;
-    comma sep;
-    int_expr intExpr;
-    end_paren endParen;
```
Comment (rmh2805) [inline src/parseTree/nodes/charAt_expr.java:15]: Why did you remove the categories for ONLY some of the nodes and not all?

File: src/parseTree/nodes/concat_expr.java [hunk 5]
```diff
@@ -1,26 +1,14 @@
...
+import java.util.ArrayList;
+import java.util.List;
 
-public class concat_expr extends str_expr implements str_val {
-    concat_label op;
-    start_paren startParen;
-    str_val lStr;
-    comma sep;
-    str_val rStr;
-    end_paren endParen;
```
Comment (rmh2805) [inline src/parseTree/nodes/concat_expr.java:15]: Inconsistent Removal

File: src/parseTree/nodes/double_expr.java [hunk 6]
```diff
@@ -9,12 +9,20 @@
...
+import java.util.List;
+
+public class double_expr extends expr<Double> implements double_val, node {
     private double_val lVal;
     private op operator;
     private double_val rVal;
+    private List<Object> children = new ArrayList<>();
 
+    public void addChild(Object child) {
+        children.add(child);
```
Comment (rmh2805) [inline src/parseTree/nodes/double_expr.java:22]: Consider defaulting this function

File: src/parseTree/nodes/print_stmt.java [hunk 7]
```diff
@@ -5,13 +5,22 @@
...
-public class print_stmt extends stmt<Integer> {
+import java.util.ArrayList;
+import java.util.List;
+
+public class print_stmt extends stmt<Integer> implements node {
     private print_label printLabel;
     private start_paren startParen;
     private expr toPrint;
     private end_paren endParen;
     private end_stmt endStmt;
```
Comment (rmh2805) [inline src/parseTree/nodes/print_stmt.java:16]: An example without removal

File: src/parser.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+public class parser {
+
+    private static boolean first(Object parent, token token) {
+        boolean result;
+        String parentName = parent.getClass().getSimpleName();
+        if (parent instanceof String) parentName = (String) parent;
+        String tokenName = token.getClass().getSimpleName();
+        if (FIRST.get(parentName) == null || FIRST.get(parentName).get(tokenName) == null) result = false;
+        else result = FIRST.get(parentName).get(tokenName);
+        return result;
```
Comment (rmh2805) [inline src/parser.java]: Wall of text formatting for 1, for another, use polymorphism to take either a string or a node along with the token. Then just pass the node through to the string form by using the getClass().getSimpleName()

**Review by rmh2805 — CHANGES_REQUESTED** [review]
There is a disturbing lack of documentation here as well, and this is pot speaking to burnt kettle here.
Also, I would suggest an explicit test module for the parser

File: src/tokenizer.java [hunk 9]
```diff
@@ -21,152 +21,115 @@
...
-                        if (!((ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9') || ch == ' '))
+                        }
+                        if (!Character.isAlphabetic(ch) && !Character.isDigit(ch) && ch != ' ')
                             errorPrinter.throwError(lineCount, new Syntax("Illegal character '" + ch +
                                     "' detected in string literal"));
-
                         tok.append(ch);
-
                         i++;
                     }
```
Comment (Jimobob) [inline src/tokenizer.java:87]: Just want to eliminate that double `i++`, remove the `i++` before while loop and at end of while loop:

```suggestion
                   while (i < lineLength) {
                        i++;
                        char ch = line.charAt(i);
```

File: src/tokenizer.java [hunk 10]
```diff
@@ -21,152 +21,115 @@
...
+                } else if (Character.isDigit(ch) || ch == '.') { // Integer or Double
                     boolean isDouble = false;
-                    int intV = -1;
-                    double dblV = -1;
-                    try {
-                        intV = Integer.parseInt(tok.toString());
-                    } catch (NumberFormatException n) {
-                        try {
-                            dblV = Double.parseDouble(tok.toString());
+                    int z = -1;
```
Comment (Jimobob) [inline src/tokenizer.java:93]: What exactly is `z`? Would prefer a more appropriate name for this var.

File: src/tokenizer.java [hunk 11]
```diff
@@ -21,152 +21,115 @@
...
+                            tokenList.add(new int_label(lineCount));
+                            break;
+                        case "String":
+                            tokenList.add(new str_label(lineCount));
+                            break;
+                        default:
+                            if (Character.isLowerCase(tok.charAt(0)))
+                                tokenList.add(new id(lineCount, tok.toString()));
+                            else errorPrinter.throwError(lineCount, new Syntax("invalid type"));
                             break;
```
Comment (Jimobob) [inline src/tokenizer.java:71]: 👍 !

**Review by Jimobob — COMMENTED** [review]
Moved all reviews from tokenizer onto this branch

File: src/parseTree/nodes/asmt.java (hunk 3 above)
Comment (rmh2805) [reply src/parseTree/nodes/asmt.java]: Must be returned to retain compatibility for execution

File: src/parseTree/nodes/node.java (hunk 2 above)
Comment (rmh2805) [reply src/parseTree/nodes/node.java]: I also suggest a `fix` method to properly assign children to their fields for execution

File: src/parseTree/nodes/double_expr.java (hunk 6 above)
Comment (rmh2805) [reply src/parseTree/nodes/double_expr.java:22]: Nevermind, I just remembered that interfaces and fields don't mix

File: src/tokenizer.java (hunk 9 above)
Comment (antxln) [reply src/tokenizer.java:87]: The `i++` before the while loop handles a str_literal that starts at the end of the line and immediately wraps around. In the suggestion, `line.charAt(i)` may throw an IndexOutOfBounds. Im sorry for the lack of documentation.

File: src/tokenizer.java (hunk 10 above)
Comment (antxln) [reply src/tokenizer.java:93]: In mathematics, Z denotes the set of integers, hence the var z

**Review by rmh2805 — APPROVED** [review]
Closed this review

File: src/parseSet.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+class parseSet {
+
+    private static final String DELIMITER = "[^a-zA-Z0-9_,]+";
+    private static final String FIRSTPATH = "Text/FIRST.txt";
+    private static final String PREDICTPATH = "Text/PREDICT.txt";
+    static Map<String, Map<String, Boolean>> FIRST = new HashMap<>();
+    static Map<String, Map<String, List<String>>> PREDICT = new HashMap<>();
+
+    /**
+     * Populate FIRST or PREDICT data structure for LL parsing given file with
```
Comment (rmh2805) [inline src/parseSet.java:16]: Why is this in its own file, as opposed to in the main body?

File: src/parseTree/nodes/asmt.java [hunk 13]
```diff
@@ -1,23 +1,36 @@
...
-        this.op = op;
-        this.exp = exp;
-        this.endStmt = endStmt;
+    public void addChild(Object child) {
+        children.add(child);
+    }
+
+    public void fixChildren() {
+        //todo Assign the proper children to their fields
+    }
```
Comment (rmh2805) [inline src/parseTree/nodes/asmt.java]: Needs to be implemented

File: src/parseTree/nodes/charAt_expr.java [hunk 14]
```diff
@@ -6,21 +6,25 @@
...
-        this.strExpr = strExpr;
-        this.sep = sep;
-        this.intExpr = intExpr;
-        this.endParen = endParen;
+    public void addChild(Object child) {
+        children.add(child);
+    }
+
+    public void fixChildren() {
+        //todo Assign the proper children to their fields
```
Comment (rmh2805) [inline src/parseTree/nodes/charAt_expr.java:27]: Needs to be implemented

File: src/parseTree/nodes/concat_expr.java [hunk 15]
```diff
@@ -6,21 +6,25 @@
...
 
-    public concat_expr(concat_label op, start_paren startParen, str_val lStr, comma sep, str_val rStr, end_paren endParen) {
-        this.op = op;
-        this.startParen = startParen;
-        this.lStr = lStr;
-        this.sep = sep;
-        this.rStr = rStr;
-        this.endParen = endParen;
+    public void fixChildren() {
+        //todo Assign the proper children to their fields
```
Comment (rmh2805) [inline src/parseTree/nodes/concat_expr.java:27]: Needs to be implemented

File: src/parseTree/nodes/double_expr.java [hunk 16]
```diff
@@ -9,12 +9,24 @@
...
     private double_val rVal;
+    private List<Object> children = new ArrayList<>();
+
+    public void addChild(Object child) {
+        children.add(child);
+    }
 
+    public void fixChildren() {
+        //todo Assign the proper children to their fields
+    }
```
Comment (rmh2805) [inline src/parseTree/nodes/double_expr.java:27]: Needs to be implemented

File: src/parseTree/nodes/asmt.java (hunk 3 above)
Comment (antxln) [reply src/parseTree/nodes/asmt.java]: I tried not to remove logic, since it wasnt necessary for parser. That being said, there was no reason to remove the fields and constructor. At the time, I was thinking the fields were unnecessary assuming the children are in the correct order (which is true if parser works correctly), but having a `fix` method and then using the fields would be more readable.

File: src/parseTree/nodes/int_expr.java [hunk 17]
```diff
@@ -9,11 +9,24 @@
...
     private int_val rVal;
+    private List<Object> children = new ArrayList<>();
+
+    public void addChild(Object child) {
+        children.add(child);
+    }
+
+    public void fixChildren() {
+        //todo Assign the proper children to their fields
+    }
```
Comment (rmh2805) [inline src/parseTree/nodes/int_expr.java:27]: Needs to be implemented

File: src/parseTree/nodes/print_stmt.java [hunk 18]
```diff
@@ -5,13 +5,26 @@
...
+    private List<Object> children = new ArrayList<>();
+
+    public void addChild(Object child) {
+        children.add(child);
+    }
+
+    public void fixChildren() {
+        //todo Assign the proper children to their fields
+    }
```
Comment (rmh2805) [inline src/parseTree/nodes/print_stmt.java:25]: Needs to be implemented

File: src/parseTree/nodes/program.java [hunk 19]
```diff
@@ -1,7 +1,19 @@
...
     private stmt_lst firstStatement;
+    private List<Object> children = new ArrayList<>();
+
+    public void addChild(Object child) {
+        children.add(child);
+    }
+
+    public void fixChildren() {
+        //todo Assign the proper children to their fields
+    }
```
Comment (rmh2805) [inline src/parseTree/nodes/program.java:16]: Needs to be implemented

File: src/parseTree/nodes/stmt_lst.java [hunk 20]
```diff
@@ -1,9 +1,22 @@
...
     private stmt_lst next;
+    private List<Object> children = new ArrayList<>();
+
+    public void addChild(Object child) {
+        children.add(child);
+    }
+
+    public void fixChildren() {
+        //todo Assign the proper children to their fields
+    }
```
Comment (rmh2805) [inline src/parseTree/nodes/stmt_lst.java:17]: Needs to be implemented

File: src/parseTree/nodes/str_literal.java [hunk 21]
```diff
@@ -4,10 +4,24 @@
...
     private quote endQuote;
+    private List<Object> children = new ArrayList<>();
+
+    public void addChild(Object child) {
+        children.add(child);
+    }
+
+    public void fixChildren() {
+        //todo Assign the proper children to their fields
+    }
```
Comment (rmh2805) [inline src/parseTree/nodes/str_literal.java:22]: Needs to be implemented

**Review by rmh2805 — CHANGES_REQUESTED** [review]
The `fixChildren` functions need to assign the abstract children to the proper fields in their classes in order to enable execution. This is an implementation of tree decoration. I suggest using recursive calls to do this as well (`program` calling it on each of the statement list elements, etc.)

File: src/tokenizer.java [hunk 22]
```diff
@@ -21,152 +21,115 @@
      */
     public static List<token> tokenize(String filepath) throws FileNotFoundException {
         Scanner sc = new Scanner(new File(filepath));
-        int lineCount = 1;  //An incrementing line counter
+        int lineCount = 0;  //An incrementing line counter
```
Comment (a1re1) [inline src/tokenizer.java:24]: should this be 0 indexed?

File: src/parseSet.java (hunk 12 above)
Comment (antxln) [reply src/parseSet.java:16]: There were alternatives for setting the parse sets so I wanted to separate the two in case we decided to change how we loaded the parse sets. Whats the benefit of having the two combined?

File: .idea/dictionaries/rmh2805.xml [hunk 23]
```diff
@@ -1,7 +0,0 @@
-<component name="ProjectDictionaryState">
```
Comment (a1re1) [inline .idea/dictionaries/rmh2805.xml:1]: not directly related, but wanted to note it here:
- we should get rid of the idea project before submitting and add a makefile or something that will handle compiling on the cs machines

File: src/parseSet.java (hunk 12 above)
Comment (a1re1) [reply src/parseSet.java:16]: i agree these set's shouldn't be in their own files. having to deal with file io and relative paths is just another avenue for failure post compiling

File: src/parseTree/nodes/expr.java [hunk 24]
```diff
@@ -1,6 +1,6 @@
 package src.parseTree.nodes;
 
-public abstract class expr<T> extends stmt<T> {
```
Comment (a1re1) [inline src/parseTree/nodes/expr.java:3]: ~~not sure why this is an abstract instead of an interface, I think we should untangle our hierarchy a bit because `expr`, `stmt`,  and `str_expr` all implement node (and it looks like most classes that extend these abstracts also implement node), and and since these are lacking concrete method definitions, it would be easier to just extend node and make these interfaces.~~

scratch that.. took another look and it can be simplified even more. I'll just propose a change that should clear this up, since describing may just get confusing

File: src/parseTree/nodes/node.java [hunk 25]
```diff
@@ -0,0 +1,15 @@
...
+public interface node {
+    /**
+     * Adds a child to the node's internal list of children
+     *
+     * @param child The child to add to the node
+     */
+    public void addChild(Object child);
+
+    /**
+     * Assign children to their fields for execution
```
Comment (a1re1) [inline src/parseTree/nodes/node.java:12]: kind of unclear on what this is supposed to do; maybe there's a better verb than fix?

File: src/parseTree/nodes/node.java [hunk 26]
```diff
@@ -0,0 +1,15 @@
+package src.parseTree.nodes;
+
+public interface node {
+    /**
+     * Adds a child to the node's internal list of children
+     *
+     * @param child The child to add to the node
+     */
+    public void addChild(Object child);
```
Comment (a1re1) [inline src/parseTree/nodes/node.java:9]: we should make node an abstract class and just implement the children list, addChild, and getChildren to DRY up all of those redundant definitions

File: src/parser.java [hunk 27]
```diff
@@ -0,0 +1,214 @@
+package src;
```
Comment (a1re1) [inline src/parser.java:1]: I know this is the meat and potatoes of this PR, but it's super hard check if this is right just by reading. I think we should work towards easier testing so that we can confidently verify this.

File: src/tokenizer.java [hunk 28]
```diff
@@ -21,152 +21,115 @@
...
         List<token> tokenList = new LinkedList<>(); //The list of tokens which will be returned
 
         while (sc.hasNextLine()) {  //For each line of the source file
+            lineCount++;
```
Comment (a1re1) [inline src/tokenizer.java:28]: jk, since we're incrementing at the beginning of the loop it makes sense that we start at 0. is there any particular reason this approach was chosen?

File: src/tokenizer.java [hunk 29]
```diff
@@ -21,152 +21,115 @@
...
                         }
-                        else {
-                            tok.append(ch);
+                        tok.append(ch);
+                        i++;
+                    }
+                    // Check reserved keywords
+                    switch (tok.toString()) {
+                        case "print":
+                            tokenList.add(new print_label(lineCount));
```
Comment (a1re1) [inline src/tokenizer.java:50]: while we're here, Johnson mentioned in class yesterday that he meant for us to throw the position in the line as well as the line number. he said the writeup that demonstrated otherwise was bad but he changed it. I'm pretty sure he said he wasn't going to penalize in this phase if we just do line number, but he would for phase 2 so he said we should change it eventually.

File: src/tokenizer.java (hunk 28 above)
Comment (antxln) [reply src/tokenizer.java:28]: When skipping comments, if the incrementer is at the end of the loop, it is skipped. Its not a big deal, difference between `if comment {count++;break;}` and `if comment {break;}`, but I see starting at index 0 is not intuitive since our output indexes starting at 1.

File: src/parseSet.java (hunk 12 above)
Comment (antxln) [reply src/parseSet.java:16]: Instead of dealing with file io, what do you propose? I think we could hardcode the HashMaps (using whats generated from parseSet) or embed the logic in each node.

File: src/parseTree/nodes/node.java (hunk 25 above)
Comment (antxln) [reply src/parseTree/nodes/node.java:12]: From my understanding, each node has a list of Objects. Based on the index, the Object is casted to a specific token/node and set to their respective fields. Maybe map or assign?

File: src/tokenizer.java (hunk 22 above)
Comment (rmh2805) [reply src/tokenizer.java:24]: No. It's not an index, instead it corresponds to the line number in the file

File: src/parseSet.java (hunk 12 above)
Comment (rmh2805) [reply src/parseSet.java:16]: Actually, It may make more sense to use the files so that we don't have to delve deep for parts 2 and 3 (or for any errors in either parse list)

File: src/tokenizer.java (hunk 10 above)
Comment (rmh2805) [reply src/tokenizer.java:93]: why not something like `intIn`?

File: src/parseSet.java (hunk 12 above)
Comment (rmh2805) [reply src/parseSet.java:16]: It also keeps the sets centralized for modification

File: src/parseSet.java [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+     */
+    static void loadParseSets() {
+        try {
+            loadSet(FIRSTPATH);
+            loadSet(PREDICTPATH);
+        } catch (FileNotFoundException e) {
+            System.out.println("Invalid file path for parse set");
+            System.exit(-1);
+        }
+    }
```
Comment (rmh2805) [inline src/parseSet.java:70]: I'd suggest combining these because, as it stands right now it doesn't remove significant amounts of copied code and starts to spaghetify

File: src/parseSet.java [hunk 31]
```diff
@@ -0,0 +1,71 @@
+package src;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.*;
+
+class parseSet {
+
+    private static final String DELIMITER = "[^a-zA-Z0-9_,]+";
```
Comment (rmh2805) [inline src/parseSet.java:9]: What are you using the delimiter to tokenize and, more to the point, is it doing so properly?

File: src/parseSet.java (hunk 30 above)
Comment (rmh2805) [reply src/parseSet.java:70]: My suggested edit is on the new branch  [parseSet_suggestion](https://github.com/rmh2805/Jott_Interpreter/tree/parseSet_suggestion)

File: src/parseSet.java (hunk 31 above)
Comment (rmh2805) [reply src/parseSet.java:9]: Never mind, I checked and I was just being a dumb bitch

**Comment by rmh2805** [comment]
Created a branch `parser_cleanup` in which I attempted to do some cleaning of the parser code and documentation as I understood it.

On going through, I do understand why you have the string literals (e.g. `"op"`) there, but I have to wonder if  it would make more sense to replace those magic literals with some `private static String`s at the top
